Data Race in activator.net.revisionWatcher

This is caused by `revisionWatcher.run()` closing `rw.destsCh` when done running, while endpoints informer still hasn't finished and calls `revisionBackendsManager.endpointsUpdated` causing a panic for sending on a closed channel.

Ideally, channels should be closed from the writing side (not from `revisionWatcher.run()` which is on receiving end) or not closed at all.

/area autoscale

## What version of Knative?

Latest master branch.

## Expected Behavior

The following should always succeed:

`$ go test -count 10 -cpu 1,2,4 -race -parallel 12 ./pkg/activator/net/...`

## Actual Behavior

Test occasionally fails with

```
==================
WARNING: DATA RACE
Write at 0x00c000779b10 by goroutine 67:
  runtime.closechan()
      /home/kaslevs/go/src/runtime/chan.go:334 +0x0
  knative.dev/serving/pkg/activator/net.(*revisionWatcher).run()
      /home/kaslevs/.go/src/knative.dev/serving/pkg/activator/net/revision_backends.go:318 +0x4ef

Previous read at 0x00c000779b10 by goroutine 70:
  runtime.chansend()
      /home/kaslevs/go/src/runtime/chan.go:142 +0x0
  knative.dev/serving/pkg/activator/net.(*revisionBackendsManager).endpointsUpdated()
      /home/kaslevs/.go/src/knative.dev/serving/pkg/activator/net/revision_backends.go:450 +0x881
  knative.dev/serving/pkg/activator/net.(*revisionBackendsManager).endpointsUpdated-fm()
      /home/kaslevs/.go/src/knative.dev/serving/pkg/activator/net/revision_backends.go:432 +0x55
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnAdd()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/controller.go:195 +0x8c
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.FilteringResourceEventHandler.OnAdd()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/controller.go:227 +0x7a
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*FilteringResourceEventHandler).OnAdd()
      <autogenerated>:1 +0x93
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1.1()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:607 +0x39e
  knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:284 +0x5e
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:601 +0xdb
  knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x6f
  knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0x108
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*processorListener).run()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0xa8
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*processorListener).run-fm()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:593 +0x41
  knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x6a

Goroutine 67 (running) created at:
  knative.dev/serving/pkg/activator/net.(*revisionBackendsManager).getOrCreateRevisionWatcher()
      /home/kaslevs/.go/src/knative.dev/serving/pkg/activator/net/revision_backends.go:423 +0x4b1
  knative.dev/serving/pkg/activator/net.(*revisionBackendsManager).endpointsUpdated()
      /home/kaslevs/.go/src/knative.dev/serving/pkg/activator/net/revision_backends.go:443 +0x22b
  knative.dev/serving/pkg/activator/net.(*revisionBackendsManager).endpointsUpdated-fm()
      /home/kaslevs/.go/src/knative.dev/serving/pkg/activator/net/revision_backends.go:432 +0x55
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnAdd()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/controller.go:195 +0x8c
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.FilteringResourceEventHandler.OnAdd()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/controller.go:227 +0x7a
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*FilteringResourceEventHandler).OnAdd()
      <autogenerated>:1 +0x93
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1.1()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:607 +0x39e
  knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:284 +0x5e
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:601 +0xdb
  knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x6f
  knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0x108
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*processorListener).run()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0xa8
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*processorListener).run-fm()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:593 +0x41
  knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x6a

Goroutine 70 (running) created at:
  knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x6f
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*sharedProcessor).addListener()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:443 +0x2fc
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*sharedIndexInformer).AddEventHandlerWithResyncPeriod()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:391 +0x2ac
  knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*sharedIndexInformer).AddEventHandler()
      /home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:327 +0x69
  knative.dev/serving/pkg/activator/net.newRevisionBackendsManagerWithProbeFrequency()
      /home/kaslevs/.go/src/knative.dev/serving/pkg/activator/net/revision_backends.go:364 +0x6b6
  knative.dev/serving/pkg/activator/net.TestRevisionBackendManagerAddEndpoint.func1()
      /home/kaslevs/.go/src/knative.dev/serving/pkg/activator/net/revision_backends_test.go:661 +0x830
  testing.tRunner()
      /home/kaslevs/go/src/testing/testing.go:909 +0x199
==================
E1024 11:57:49.337253   31556 runtime.go:73] Observed a panic: "send on closed channel" (send on closed channel)
goroutine 7127 [running]:
knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x20b4500, 0x263c250)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:69 +0x8c
knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51 +0xbe
panic(0x20b4500, 0x263c250)
	/home/kaslevs/go/src/runtime/panic.go:679 +0x1b2
knative.dev/serving/pkg/activator/net.(*revisionBackendsManager).endpointsUpdated(0xc00073b900, 0x22d1160, 0xc000745e00)
	/home/kaslevs/.go/src/knative.dev/serving/pkg/activator/net/revision_backends.go:450 +0x882
knative.dev/serving/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(...)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/controller.go:195
knative.dev/serving/vendor/k8s.io/client-go/tools/cache.FilteringResourceEventHandler.OnAdd(0xc000827540, 0x26a1fa0, 0xc000827560, 0x22d1160, 0xc000745e00)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/controller.go:227 +0x7b
knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0x1055d4c, 0x349001901ee, 0x7f6704bc25a0)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:607 +0x39f
knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0, 0xc00069bdc8, 0xc000083f28, 0x1edacd1)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:284 +0x5f
knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:601 +0xdc
knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000083f38)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x70
knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00069bf38, 0xdf8475800, 0x0, 0xc000083f01, 0xc000779aa0)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0x109
knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.Until(...)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*processorListener).run(0xc00073b980)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:599 +0xa9
knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc000112760, 0xc0004ddeb0)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x6b
created by knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x70
panic: send on closed channel [recovered]
	panic: send on closed channel

goroutine 7127 [running]:
knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x169
panic(0x20b4500, 0x263c250)
	/home/kaslevs/go/src/runtime/panic.go:679 +0x1b2
knative.dev/serving/pkg/activator/net.(*revisionBackendsManager).endpointsUpdated(0xc00073b900, 0x22d1160, 0xc000745e00)
	/home/kaslevs/.go/src/knative.dev/serving/pkg/activator/net/revision_backends.go:450 +0x882
knative.dev/serving/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(...)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/controller.go:195
knative.dev/serving/vendor/k8s.io/client-go/tools/cache.FilteringResourceEventHandler.OnAdd(0xc000827540, 0x26a1fa0, 0xc000827560, 0x22d1160, 0xc000745e00)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/controller.go:227 +0x7b
knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0x1055d4c, 0x349001901ee, 0x7f6704bc25a0)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:607 +0x39f
knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0, 0xc00069bdc8, 0xc000083f28, 0x1edacd1)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:284 +0x5f
knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:601 +0xdc
knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000083f38)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x70
knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00069bf38, 0xdf8475800, 0x0, 0xc000083f01, 0xc000779aa0)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0x109
knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.Until(...)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
knative.dev/serving/vendor/k8s.io/client-go/tools/cache.(*processorListener).run(0xc00073b980)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/client-go/tools/cache/shared_informer.go:599 +0xa9
knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc000112760, 0xc0004ddeb0)
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x6b
created by knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/home/kaslevs/.go/src/knative.dev/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x70
FAIL	knative.dev/serving/pkg/activator/net	81.164s
FAIL
```

--- a/pkg/activator/net/revision_backends.go
+++ b/pkg/activator/net/revision_backends.go
@@ -68,11 +68,12 @@ const (
 // revisionWatcher watches the podIPs and ClusterIP of the service for a revision. It implements the logic
 // to supply revisionDestsUpdate events on updateCh
 type revisionWatcher struct {
-	doneCh   <-chan struct{}
+	stopCh   <-chan struct{}
 	cancel   context.CancelFunc
 	rev      types.NamespacedName
 	protocol networking.ProtocolType
 	updateCh chan<- revisionDestsUpdate
+	done     chan struct{}
 
 	// Stores the list of pods that have been successfully probed.
 	healthyPods sets.String
@@ -95,11 +96,12 @@ func newRevisionWatcher(ctx context.Context, rev types.NamespacedName, protocol 
 	logger *zap.SugaredLogger) *revisionWatcher {
 	ctx, cancel := context.WithCancel(ctx)
 	return &revisionWatcher{
-		doneCh:          ctx.Done(),
+		stopCh:          ctx.Done(),
 		cancel:          cancel,
 		rev:             rev,
 		protocol:        protocol,
 		updateCh:        updateCh,
+		done:            make(chan struct{}),
 		healthyPods:     sets.NewString(),
 		transport:       transport,
 		destsCh:         destsCh,
@@ -218,7 +220,7 @@ func (rw *revisionWatcher) probePodIPs(dests sets.String) (sets.String, bool, er
 
 func (rw *revisionWatcher) sendUpdate(clusterIP string, dests sets.String) {
 	select {
-	case <-rw.doneCh:
+	case <-rw.stopCh:
 		return
 	default:
 		rw.updateCh <- revisionDestsUpdate{Rev: rw.rev, ClusterIPDest: clusterIP, Dests: dests}
@@ -293,7 +295,7 @@ func (rw *revisionWatcher) checkDests(dests sets.String) {
 }
 
 func (rw *revisionWatcher) run(probeFrequency time.Duration) {
-	defer close(rw.destsCh)
+	defer close(rw.done)
 
 	var dests sets.String
 	timer := time.NewTicker(probeFrequency)
@@ -314,7 +316,7 @@ func (rw *revisionWatcher) run(probeFrequency time.Duration) {
 		}
 
 		select {
-		case <-rw.doneCh:
+		case <-rw.stopCh:
 			return
 		case x := <-rw.destsCh:
 			dests = x
@@ -386,7 +388,7 @@ func newRevisionBackendsManagerWithProbeFrequency(ctx context.Context, tr http.R
 		rbm.revisionWatchersMux.Lock()
 		defer rbm.revisionWatchersMux.Unlock()
 		for _, rw := range rbm.revisionWatchers {
-			<-rw.destsCh
+			<-rw.done
 		}
 	}()
 
@@ -447,7 +449,11 @@ func (rbm *revisionBackendsManager) endpointsUpdated(newObj interface{}) {
 	}
 	dests := endpointsToDests(endpoints, networking.ServicePortName(rw.protocol))
 	rbm.logger.Debugf("Updating Endpoints: %q (backends: %d)", revID.String(), len(dests))
-	rw.destsCh <- dests
+	select {
+	case <-rbm.ctx.Done():
+		return
+	case rw.destsCh <- dests:
+	}
 }
 
 // deleteRevisionWatcher deletes the revision watcher for rev if it exists. It expects

--- a/pkg/activator/net/revision_backends_test.go
+++ b/pkg/activator/net/revision_backends_test.go
@@ -407,19 +407,19 @@ func TestRevisionWatcher(t *testing.T) {
 				t.Errorf("revisionDests updates = %v, want: %v, diff (-want, +got):\n %s", got, want, cmp.Diff(want, got))
 			}
 
-			assertChClosed(t, destsCh)
+			assertChClosed(t, rw.done)
 		})
 	}
 }
 
-func assertChClosed(t *testing.T, ch chan sets.String) {
+func assertChClosed(t *testing.T, ch chan struct{}) {
 	defer func() {
 		if r := recover(); r == nil {
 			t.Errorf("the channel was not closed")
 		}
 	}()
 	select {
-	case ch <- nil:
+	case ch <- struct{}{}:
 		// Panics if the channel is closed
 	default:
 		// Prevents from blocking forever if the channel is not closed
@@ -719,7 +719,7 @@ func TestCheckDests(t *testing.T) {
 		updateCh:         uCh,
 		serviceLister:    si.Lister(),
 		logger:           TestLogger(t),
-		doneCh:           dCh,
+		stopCh:           dCh,
 	}
 	rw.checkDests(sets.NewString("10.1.1.5"))
 	select {
@@ -800,7 +800,7 @@ func TestCheckDestsSwinging(t *testing.T) {
 		updateCh:        uCh,
 		serviceLister:   si.Lister(),
 		logger:          TestLogger(t),
-		doneCh:          dCh,
+		stopCh:          dCh,
 		podsAddressable: true,
 		transport:       network.RoundTripperFunc(fakeRT.RT),
 	}

